File: py/day02.py
rock = "R"
paper = "P"
scissors = "S"

win = "W"
lose = "L"
draw = "D"

move = {
    "A": rock,
    "B": paper,
    "C": scissors,
    "X": rock,
    "Y": paper,
    "Z": scissors
}

op_moves = {
    rock: "A",
    paper: "B",
    scissors: "C"
}

my_moves = {
    rock: "X",
    paper: "Y",
    scissors: "Z"
}

my_winning_moves = {
    rock: my_moves[paper],
    paper: my_moves[scissors],
    scissors: my_moves[rock]
}

my_losing_moves = {
    rock: my_moves[scissors],
    paper: my_moves[rock],
    scissors: my_moves[paper]
}

strategy = {
    "X": lose, 
    "Y": draw, 
    "Z": win
}

move_score = {
    rock: 1,
    paper: 2,
    scissors: 3
}

def evaluate_game(op, me):
    if move[op] == rock:
        if move[me] == rock:
            return draw
        elif move[me] == paper:
            return win
        else:
            return lose
    elif move[op] == paper:
        if move[me] == rock:
            return lose
        elif move[me] == paper:
            return draw
        else:
            return win
    else:
        if move[me] == rock:
            return win
        elif move[me] == paper:
            return lose
        else:
            return draw

def evaluate_strategy(o, s):
    the_strategy = strategy[s]
    if the_strategy == win:
        return my_winning_moves[move[o]]
    elif the_strategy == lose:
        return my_losing_moves[move[o]]
    else:
        return my_moves[move[o]]

def evaluate_score(op, me):
    score = 0
    result = evaluate_game(op, me)
    if result == win:
        score += 6
    elif result == draw:
        score += 3
    score += move_score[move[me]]
    return score

def main():
    f = open("../input/02/input.txt", "r")

    score = 0
    for line in f.readlines():
        moves = line.strip().split(" ")
        op = moves[0]
        # me = moves[1]
        me = evaluate_strategy(op, moves[1])
        score += evaluate_score(op, me)
    print(score)


if __name__ == "__main__":
    main()

File: hs/src/Day02.hs
module Day02 (run) where

opRock :: Char
opRock = 'A'

opPaper :: Char
opPaper = 'B'

opScissors :: Char
opScissors = 'C'

meRock :: Char
meRock = 'X'

mePaper :: Char
mePaper = 'Y'

meScissors :: Char
meScissors = 'Z'

strategyLose :: Char
strategyLose = 'X'

strategyDraw :: Char
strategyDraw = 'Y'

strategyWin :: Char
strategyWin = 'Z'

winScore :: Int
winScore = 6

drawScore :: Int
drawScore = 3

run :: IO ()
run = do
    contents <- readFile "../input/02/input.txt"
    print . scoreGames
          . lines $ contents

scoreGames :: [String] -> Int
scoreGames xs = sg xs 0
    where
        sg ((op:' ':strategy:[]):rest) score = do
            let me = evaluateStrategy op strategy
            let result = evaluateGame op me
            sg rest $ (score + result + (moveScore me))
        sg [] score = score

evaluateGame :: Char -> Char -> Int
evaluateGame op me
    | op == opRock && me == meRock = drawScore
    | op == opPaper && me == mePaper = drawScore
    | op == opScissors && me == meScissors = drawScore
    | op == opRock && me == mePaper = winScore
    | op == opPaper && me == meScissors = winScore
    | op == opScissors && me == meRock = winScore
    | otherwise = 0

moveScore :: Char -> Int
moveScore me
    | me == meRock = 1
    | me == mePaper = 2
    | me == meScissors = 3
    | otherwise = error "Invalid move"

evaluateStrategy :: Char -> Char -> Char
evaluateStrategy op strategy
    | strategy == strategyWin = evaluateWin op
    | strategy == strategyDraw = evaluateDraw op
    | strategy == strategyLose = evaluateLose op
    | otherwise = error "Invalid strategy"
        where
            evaluateWin strategyOp
                | strategyOp == opRock = mePaper
                | strategyOp == opPaper = meScissors
                | strategyOp == opScissors = meRock
                | otherwise = error "Invalid move"
            evaluateDraw strategyOp
                | strategyOp == opRock = meRock
                | strategyOp == opPaper = mePaper
                | strategyOp == opScissors = meScissors
                | otherwise = error "Invalid move"
            evaluateLose strategyOp
                | strategyOp == opRock = meScissors
                | strategyOp == opPaper = meRock
                | strategyOp == opScissors = mePaper
                | otherwise = error "Invalid move"
